### DISCUSSION TITLE
Workflow Failure Recovery

### DISCUSSION THREAD

**Discussion by dennissetiawan** [indeedeng/iwf#310] [Q&A]
Hello , i have cases where I limit the duration/attemp of a state execution and I want to do rollback on my data in case of workflow failure to fix the data. Like a defer / recovery function in golang of some sort? Can we do this in iwf / what is the recommendation to do this?

**Comment by longquanzheng** [comment]
A common pattern that we are using is:
- put the logic in waitUntil with WaitUntilApiFailurePolicy.PROCEED_ON_FAILURE . It can just return empty commandRequest after the logic. 
- This failure policy (in state options) will let the state continue to execute api after waitUntil has maxed out attempt/duration, rather than failing the workflow 
- If waitUntil succeed, then continue the workflow as normal
- If waitUntil fails, then do a cleanup/recovery as you need before failing the workflow . 

To check if waitUntil has succeeded, there is a field in CommandResults.getWaitUntilApiSucceeded()

**Reply by longquanzheng** [reply]
See the docs https://github.com/indeedeng/iwf#waituntil-api-failure-policy

**Reply by dennissetiawan** [reply]
Okay, so we can execute the recovery code in Execute API after set the policy if the GetWaitUntilApiSuceeded return false right?
when does that returns nil (in go SDK) `StateWaitUntilApiSucceeded *bool`.

Also what if the code i want to recover from executes in the Execute() API function?
Is there a way or i should just move it to another state?

**Reply by longquanzheng** [reply]
It will not return nil unless your server is very old. This was for a compatibility for the very old sever

**Reply by longquanzheng** [reply]
Yes another way is to utilize the “attempts” or “firstAttemptStartTime” from the context, and make sure it don’t use up the attempts , and return a decision to another state to cleanup/recover

**Reply by longquanzheng** [reply]
For execute api, ideally we can do this
https://github.com/indeedeng/iwf/issues/311

But it’s not prioritized yet . We are working on adding more resources to this project hopefully this will come later
